pkg/common/glob/glob.go

```diff
@@ -0,0 +1,123 @@
+package common
+
+import (
+	"fmt"
+
+	"github.com/gobwas/glob"
+)
+
+// Filter is a generic filter for excluding and including globs (limited
+// regular expressions). Exclusion takes precedence if both include and exclude
+// lists are provided.
+type Filter struct {
+	exclude []glob.Glob
+	include []glob.Glob
+}
+
+type globFilterOpt func(*Filter) error
+
+// WithExcludeGlobs adds exclude globs to the filter.
+func WithExcludeGlobs(excludes ...string) globFilterOpt {
+	return func(f *Filter) error {
+		for _, exclude := range excludes {
+			g, err := glob.Compile(exclude)
+			if err != nil {
+				return err
+			}
+			f.exclude = append(f.exclude, g)
+		}
+		return nil
+	}
+}
+
+// WithIncludeGlobs adds include globs to the filter.
+func WithIncludeGlobs(includes ...string) globFilterOpt {
+	return func(f *Filter) error {
+		for _, include := range includes {
+			g, err := glob.Compile(include)
+			if err != nil {
+				return err
+			}
+			f.include = append(f.include, g)
+		}
+		return nil
+	}
+}
+
+// NewGlobFilter creates a new Filter with the provided options.
+func NewGlobFilter(opts ...globFilterOpt) (*Filter, error) {
+	filter := &Filter{}
+	for _, opt := range opts {
+		if err := opt(filter); err != nil {
+			return nil, err
+		}
+	}
+	return filter, nil
+}
+
+// Pass returns whether the object is not in the include list or in the exclude
+// list.
+func (f *Filter) Pass(object string) bool {
+	if f == nil {
```

Comment:
i didn't even know that go let you call methods on a nil receiver!

Reply:
Yes! I usually think of method receivers as syntax sugar for `func Pass(f *Filter, object string) bool`. I think it's possible because `nil` is typed in Go.

It's not often I rely on that functionality, but I thought it would be a nice touch here.

Reply:
fun trivia fact: this capability was in .net framework 1 (way back in the day) and isn't precluded by the implementation but the designers intentionally forbade it because it confused too many people